File: tests/python/realworld_deserialize.py
"""
Real-world insecure deserialization test cases for Kunlun-M scanner.
Simulates a Flask application with session/cache/import endpoints.
"""

import json
import pickle
from flask import Blueprint, request, jsonify

import_bp = Blueprint('import', __name__)


# ---------------------------------------------------------------------------
# VULN 1: pickle.loads with user-controlled serialized data
# ---------------------------------------------------------------------------
@import_bp.route('/api/import/session', methods=['POST'])
def import_session():
    """Import a user session from serialized data."""
    serialized_data = request.data
    # VULN: User-controlled data deserialized with pickle — allows arbitrary code execution
    # Attacker can craft a malicious pickle payload that executes os.system("id")
    try:
        session_data = pickle.loads(serialized_data)
    except Exception as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'imported': str(session_data)})


# ---------------------------------------------------------------------------
# VULN 2: pickle.load from user-uploaded file
# ---------------------------------------------------------------------------
@import_bp.route('/api/import/cache', methods=['POST'])
def import_cache():
    """Import cache data from uploaded file."""
    uploaded_file = request.files.get('cache_file')
    if not uploaded_file:
        return jsonify({'error': 'No file uploaded'}), 400
    # VULN: User-uploaded file contents deserialized with pickle
    # Attacker uploads a file containing a malicious pickle payload
    file_content = uploaded_file.read()
    try:
        cache_data = pickle.loads(file_content)
    except Exception as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'keys': list(cache_data.keys())})


# ---------------------------------------------------------------------------
# VULN 3: pickle.loads with base64-encoded user data
# ---------------------------------------------------------------------------
@import_bp.route('/api/import/settings', methods=['POST'])
def import_settings():
    """Import application settings from a base64-encoded pickle blob."""
    import base64
    encoded = request.form.get('settings', '')
    # VULN: Base64-decoded user input fed directly into pickle.loads
    # Attacker base64-encodes a malicious pickle and submits it
    raw_data = base64.b64decode(encoded)
    try:
        settings = pickle.loads(raw_data)
    except Exception as e:
        return jsonify({'error': str(e)}), 400
    return jsonify({'settings': settings})


# ---------------------------------------------------------------------------
# SAFE 1: json.loads for structured data import
# ---------------------------------------------------------------------------
@import_bp.route('/api/import/config', methods=['POST'])
def import_config():
    """Import application configuration from JSON."""
    json_data = request.data
    # SAFE: json.loads only parses data structures — cannot execute code
    try:
        config = json.loads(json_data)
    except json.JSONDecodeError as e:
        return jsonify({'error': f'Invalid JSON: {e}'}), 400
    return jsonify({'config': config})


# ---------------------------------------------------------------------------
# SAFE 2: pickle.loads with fixed/internally-generated data only
# ---------------------------------------------------------------------------
@import_bp.route('/api/export/state', methods=['GET'])
def export_state():
    """Export current application state (internally generated)."""
    # SAFE: Data is generated server-side, not from user input
    state = {'version': '1.0', 'items': [1, 2, 3]}
    serialized = pickle.dumps(state)
    return jsonify({'data': serialized.hex()})

File: tests/nodejs/realworld_deserialize.js
/**
 * Real-world Unsafe Deserialization test cases for Kunlun-M scanner
 * Tests: node-serialize, node-v8-serialize unserialize() with user-controlled data
 */

const express = require('express');
const serialize = require('node-serialize');
const router = express.Router();

router.use(express.json());
router.use(express.urlencoded({ extended: true }));

/**
 * VULN: Unsafe deserialization via node-serialize.unserialize with req.body
 * Attacker can craft a serialized object that executes arbitrary code when deserialized
 * node-serialize's unserialize() can execute __defineGetter__ payloads
 * Example POST: {"data": '{"___class":"Function","___args":"","___code":"return require(\\\"child_process\\\").execSync(\\\"id\\\")"}'}
 */
router.post('/session-restore', (req, res) => {
    const sessionData = req.body.data;
    try {
        // VULN: Unserialize user-controlled data - allows RCE via crafted payloads
        const session = serialize.unserialize(sessionData);
        res.json({ session });
    } catch (err) {
        res.status(400).send('Invalid session data');
    }
});

/**
 * VULN: Unsafe deserialization via node-serialize with req.query token
 * Deserializing a base64-encoded serialized object from query params
 * Example: /auth?token=<base64-encoded malicious serialized object>
 */
router.get('/auth', (req, res) => {
    const token = req.query.token;
    try {
        // VULN: Deserializing user-supplied token without validation
        const decoded = Buffer.from(token, 'base64').toString();
        const session = serialize.unserialize(decoded);
        if (session && session.userId) {
            res.json({ authenticated: true, user: session.userId });
        } else {
            res.status(401).send('Invalid token');
        }
    } catch (err) {
        res.status(401).send('Authentication failed');
    }
});

/**
 * VULN: Unsafe deserialization via node-serialize with req.params
 * Attacker can inject via URL parameters
 * Example: /restore/<base64 encoded malicious serialized payload>
 */
router.get('/restore/:data', (req, res) => {
    const data = req.params.data;
    try {
        // VULN: Deserializing URL parameter directly
        const obj = serialize.unserialize(data);
        res.json(obj);
    } catch (err) {
        res.status(400).send('Restore failed');
    }
});

/**
 * SAFE: Using JSON.parse instead of unserialize
 * JSON.parse only handles primitive types and plain objects - no code execution possible
 */
router.post('/safe-session', (req, res) => {
    const sessionData = req.body.data;
    try {
        // SAFE: JSON.parse only handles data types, cannot execute code
        const session = JSON.parse(sessionData);
        // Additional validation: ensure it's a plain object with expected fields
        if (typeof session !== 'object' || session === null || Array.isArray(session)) {
            return res.status(400).send('Invalid session format');
        }
        if (typeof session.userId !== 'string' || typeof session.role !== 'string') {
            return res.status(400).send('Missing required session fields');
        }
        res.json({ session });
    } catch (err) {
        res.status(400).send('Invalid session data');
    }
});

/**
 * SAFE: Using JSON.parse with schema validation
 * Validate the parsed object structure before using it
 */
router.get('/safe-auth', (req, res) => {
    const token = req.query.token;
    try {
        // SAFE: Parse as JSON (not serialized object), then validate structure
        const decoded = Buffer.from(token, 'base64').toString();
        const session = JSON.parse(decoded);
        // Strict type checking prevents prototype pollution and unexpected objects
        const allowedRoles = ['user', 'admin', 'viewer'];
        if (typeof session.userId !== 'number' ||
            typeof session.expiresAt !== 'number' ||
            !allowedRoles.includes(session.role)) {
            return res.status(401).send('Invalid token structure');
        }
        if (Date.now() > session.expiresAt) {
            return res.status(401).send('Token expired');
        }
        res.json({ authenticated: true, user: session.userId, role: session.role });
    } catch (err) {
        res.status(401).send('Authentication failed');
    }
});

module.exports = router;
